File: packages/jev-router/loop-stop.mjs
/**
 * Loop-stop / escalate envelope (Suraj #4 + #10).
 * Jev answers continue|stop|escalate; code maps to gate / exec / HITL.
 * Escalate is human-in-the-loop — not auto-retry, not auto-promote.
 */

import { LOOP_STOP_POLICY, applyLoopStopThresholds } from "./policy.mjs";

export function clipDigest(text, policy = LOOP_STOP_POLICY) {
  const max = Number(policy.maxDigestChars || 4000);
  return String(text || "").slice(0, max);
}

/**
 * Shadow always execs. Active: stop/escalate do not exec; continue/auto execs.
 */
export function shouldExecAgent({ mode, choice, gate, blocked } = {}) {
  const m = String(mode || "shadow").toLowerCase();
  if (m !== "active") return true;
  const c = String(choice || "");
  const g = String(gate || "");
  const b = blocked === true || blocked === "true";
  if (c === "stop" || c === "escalate") return false;
  if (b) return false;
  return c === "continue" || g === "auto";
}

/**
 * Map a loop-stop Choice answer + mode onto the router JSON contract.
 * Active + uncertain → explicit escalate (HITL, fail closed).
 * Shadow + uncertain → unclassified, does not block.
 */
export function decideLoopStop(answer, { mode = "shadow", intent = "", stepDigest = "" } = {}, policy = LOOP_STOP_POLICY) {
  const decision = applyLoopStopThresholds(answer, policy);
  const m = String(mode || "shadow").toLowerCase();
  const digest = clipDigest(stepDigest, policy);

  let choice;
  let gate;
  let reason = decision.reason;

  if (decision.outcome === "continue") {
    choice = "continue";
    gate = "auto";
  } else if (decision.outcome === "stop") {
    choice = "stop";
    gate = "hold";
  } else if (decision.outcome === "escalate") {
    choice = "escalate";
    gate = "hold";
  } else if (m === "active") {
    choice = "escalate";
    gate = "hold";
    reason = decision.reason === "cannot_tell" ? "cannot_tell" : "model_uncertain";
  } else {
    choice = "unclassified";
    gate = "auto";
  }

  const hitl = choice === "escalate";
  const blocked = m === "active" && (choice === "stop" || choice === "escalate" || gate !== "auto");
  const exec = shouldExecAgent({ mode: m, choice, gate, blocked });

  return {
    choice,
    gate,
    blocked,
    exec,
    hitl,
    autoRetry: false,
    autoPromote: false,
    intent,
    stepDigest: digest,
    loop: {
      policy: policy.version,
      selected: decision.selected,
      outcome: decision.outcome,
      reason,
      confidence: decision.confidence,
      selectedProb: decision.selectedProb,
      margin: decision.margin,
      hitl,
      autoRetry: false,
      autoPromote: false,
    },
  };
}

export function missingKeyLoop(mode, intent = "", stepDigest = "") {
  const m = String(mode || "shadow").toLowerCase();
  if (m === "active") {
    const decided = decideLoopStop({ choice: "cannot_tell" }, { mode: m, intent, stepDigest });
    decided.loop = { ...decided.loop, reason: "missing_key" };
    return decided;
  }
  const digest = clipDigest(stepDigest);
  return {
    choice: "unclassified",
    gate: "auto",
    blocked: false,
    exec: true,
    hitl: false,
    autoRetry: false,
    autoPromote: false,
    intent,
    stepDigest: digest,
    loop: {
      policy: LOOP_STOP_POLICY.version,
      selected: "unclassified",
      outcome: "cannot_tell",
      reason: "missing_key",
      hitl: false,
      autoRetry: false,
      autoPromote: false,
      shadow: true,
      blocked: false,
    },
  };
}

File: crates/bezel-bridle/src/loop_stop.rs
//! Loop-stop thresholds. Escalate stays human-in-the-loop.
//! A sub-threshold continue is `cannot_tell`, not a continue.

use crate::policy::{LOOP_STOP_MIN_CONFIDENCE, LOOP_STOP_MIN_MARGIN, LOOP_STOP_MIN_PROBABILITY};

#[derive(Debug, Clone, PartialEq)]
pub struct LoopStopDecision {
    pub outcome: &'static str,
    pub reason: &'static str,
}

pub fn apply_loop_stop_thresholds(choice: &str, confidence: f64, probabilities: &[(&str, f64)]) -> LoopStopDecision {
    let selected_prob = probabilities
        .iter()
        .find(|(label, _)| *label == choice)
        .map(|(_, value)| *value)
        .unwrap_or(0.0);
    let runner_up = probabilities
        .iter()
        .filter(|(label, _)| *label != choice)
        .map(|(_, value)| *value)
        .fold(0.0_f64, f64::max);
    let margin = selected_prob - runner_up;
    let confident = confidence >= LOOP_STOP_MIN_CONFIDENCE
        && selected_prob >= LOOP_STOP_MIN_PROBABILITY
        && margin >= LOOP_STOP_MIN_MARGIN;

    if choice == "escalate" {
        return LoopStopDecision {
            outcome: "escalate",
            reason: if confident { "selected" } else { "escalate_uncertain" },
        };
    }
    if choice != "continue" && choice != "stop" {
        return LoopStopDecision {
            outcome: "cannot_tell",
            reason: "cannot_tell",
        };
    }
    if !confident {
        return LoopStopDecision {
            outcome: "cannot_tell",
            reason: "model_uncertain",
        };
    }
    LoopStopDecision {
        outcome: if choice == "continue" { "continue" } else { "stop" },
        reason: "selected",
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct LoopEnvelope {
    pub choice: &'static str,
    pub gate: &'static str,
    pub blocked: bool,
    pub exec: bool,
    pub hitl: bool,
    pub auto_retry: bool,
    pub auto_promote: bool,
    pub intent: String,
    pub step_digest: String,
    pub loop_selected: String,
    pub loop_outcome: &'static str,
    pub loop_reason: &'static str,
    pub loop_hitl: bool,
}

pub fn decide_loop_stop(
    choice: &str,
    confidence: f64,
    probabilities: &[(&str, f64)],
    mode: &str,
    intent: &str,
    step_digest: &str,
) -> LoopEnvelope {
    let decision = apply_loop_stop_thresholds(choice, confidence, probabilities);
    let active = mode.eq_ignore_ascii_case("active");
    let (choice_word, gate, reason) = if decision.outcome == "continue" {
        ("continue", "auto", decision.reason)
    } else if decision.outcome == "stop" {
        ("stop", "hold", decision.reason)
    } else if decision.outcome == "escalate" {
        ("escalate", "hold", decision.reason)
    } else if active {
        (
            "escalate",
            "hold",
            if decision.reason == "cannot_tell" {
                "cannot_tell"
            } else {
                "model_uncertain"
            },
        )
    } else {
        ("unclassified", "auto", decision.reason)
    };
    let hitl = choice_word == "escalate";
    let blocked = active && (choice_word == "stop" || choice_word == "escalate" || gate != "auto");
    let exec = crate::policy::should_exec_agent(if active { "active" } else { "shadow" }, choice_word, gate, blocked);
    LoopEnvelope {
        choice: choice_word,
        gate,
        blocked,
        exec,
        hitl,
        auto_retry: false,
        auto_promote: false,
        intent: intent.to_string(),
        step_digest: crate::facts::clip_to_max_hunk_chars(step_digest, Some(crate::policy::MAX_DIGEST_CHARS)),
        loop_selected: decision.outcome.to_string(),
        loop_outcome: decision.outcome,
        loop_reason: reason,
        loop_hitl: hitl,
    }
}

pub fn missing_key_loop(mode: &str, intent: &str, step_digest: &str) -> LoopEnvelope {
    if mode.eq_ignore_ascii_case("active") {
        let mut decided = decide_loop_stop("cannot_tell", 0.0, &[], "active", intent, step_digest);
        decided.loop_reason = "missing_key";
        return decided;
    }
    LoopEnvelope {
        choice: "unclassified",
        gate: "auto",
        blocked: false,
        exec: true,
        hitl: false,
        auto_retry: false,
        auto_promote: false,
        intent: intent.to_string(),
        step_digest: crate::facts::clip_to_max_hunk_chars(step_digest, Some(crate::policy::MAX_DIGEST_CHARS)),
        loop_selected: "unclassified".to_string(),
        loop_outcome: "cannot_tell",
        loop_reason: "missing_key",
        loop_hitl: false,
    }
}
